Improve checkpointing for Zero stage 1

ashbhandare opened the pull request:
This PR does the folllowing changes:
1) Completely shard the FP32 weight in case of an fp16 run
2) Simplify the aggregation logic for Zero checkpoints

The correctness has been verified with the test added through https://github.com/microsoft/onnxruntime/pull/5476:
`CUDA_VISIBLE_DEVICES=0,1,2,3 mpirun -n 4 --tag-output python orttraining_run_bert_pretrain.py ORTBertPretrainTest.test_pretrain_zero`

thiagocrepaldi commented on orttraining/orttraining/python/training/checkpoint.py:
```diff
@@ -168,114 +183,80 @@ def __init__(self, checkpoint_files, clean_state_dict=None):
             if 'model' in rank_state_dict:
                 rank_state_dict = rank_state_dict['model']
-
+            
```
Check code style for the python files otherwise `flake8` will raise a failure in the CI

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:
```diff
@@ -317,6 +335,14 @@ def setup_training(args):
     return device, args
 
+def setup_torch_distributed(world_rank, world_size):
+    os.environ['RANK'] = str(world_rank)
+    os.environ['WORLD_SIZE'] = str(world_size)
```
Why do we need rank and size as env var? `ORTTrainer` reads these information from `ORTTrainerOptions.distributed` sub-dict

ashbhandare replied:
need to set it to use torch distributed barrier

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:
```diff
@@ -317,6 +335,14 @@ def setup_training(args):
+    os.environ['WORLD_SIZE'] = str(world_size)
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = str('12345')
+    torch.distributed.init_process_group(backend='nccl', world_size=world_size,
```
Is `torch.distributed` used for anything else than creating barriers to coordinate ranks for checkpointing API?

ashbhandare replied:
It is only used for creating barriers so that a single test can execute the entire workflow.

thiagocrepaldi commented:
> This PR does the folllowing changes:
> 
>     1. Completely shard the FP32 weight in case of an fp16 run
> 
>     2. Simplify the aggregation logic for Zero checkpoints
> 
> 
> The correctness has been verified with the test added through #5476:
> `CUDA_VISIBLE_DEVICES=0,1,2,3 mpirun -n 4 --tag-output python orttraining_run_bert_pretrain.py ORTBertPretrainTest.test_pretrain_zero`

Is this PR enabling any new scenario for checkpointing, such as fp32 -> fp16 and vice-versa? Or this is just a performance improvement for the existing scenarios

thiagocrepaldi commented:
@baijumeswani FYI

jessebenson commented on orttraining/orttraining/core/session/training_session.cc:
```diff
@@ -1000,6 +1004,9 @@ std::unordered_set<std::string> TrainingSession::GetStateTensorNames() const {
   std::unordered_set<std::string> checkpointed_tensor_names{};
   checkpointed_tensor_names.insert(
       weights_to_train_.begin(), weights_to_train_.end());
+  for (auto& p : updated_weight_names_map_) {
```
Is there a way to update the `weights_to_train_` with the new weight names, rather than keeping around both structures?

ashbhandare replied:
We are working on restructuring how we maintain the mapping between the original weight names passed and the modifications ORT does( fp16 weight name, associated optimizer state names, modified fp32 weight name if any) so that we can pass enough information to reconstruct the state in checkpointing without relying on string matching. This will be changed in that task.

jessebenson commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,94 @@ static Status AddL2NormNcclAllReduce(
+        initializer_partition.add_dims(partition_size);
+        initializer_partition.set_raw_data(initializer_data + partition_offset, partition_size * sizeof(float));
+
+        //Replace the old initializer with the new one
+        graph.RemoveInitializedTensor(initializer_name);
+        graph_utils::AddInitializer(graph, initializer_partition);
```
Nice

jessebenson commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,94 @@ static Status AddL2NormNcclAllReduce(
+    const std::string& initializer_name,
+    const std::vector<TensorShape>& shapes,
+    std::unordered_map<std::string, std::string>& updated_weight_names_map){
+  int64_t partition_offset = shapes[0].GetDims()[0];
+  int64_t partition_size = shapes[1].GetDims()[0];
```
Consider passing these as parameters explicitly, instead of packing into a vector of shapes.

ashbhandare replied:
will change

ashbhandare commented:
> > This PR does the folllowing changes:
> > ```
> > 1. Completely shard the FP32 weight in case of an fp16 run
> > 
> > 2. Simplify the aggregation logic for Zero checkpoints
> > ```
> > 
> > 
> > The correctness has been verified with the test added through #5476:
> > `CUDA_VISIBLE_DEVICES=0,1,2,3 mpirun -n 4 --tag-output python orttraining_run_bert_pretrain.py ORTBertPretrainTest.test_pretrain_zero`
> 
> Is this PR enabling any new scenario for checkpointing, such as fp32 -> fp16 and vice-versa? Or this is just a performance improvement for the existing scenarios

The fp16-> fp32 support already exists. This PR removes dependency of zero aggregation on optimizer state being present in the state_dict. This allows only saving the model weights for zero_1.

ashbhandare commented:
The test graph bert_toy_postprocessed.onnx had to be run through onnxruntime/tools/python/remove_initializer_from_input.py to move the initializers bert.embeddings.position_embeddings.weight and bert.embeddings.word_embeddings.weight from inputs as trainable weights should not be expected to be overridden, and zero partitioning is conditioned upon the initializers not being in graph inputs. 

Additionally, the script orttraining/orttraining/test/python/orttrainer_bert_toy_onnx_ckpt_gen.py has been added to generate the zero checkpoints required for test testToyBertCheckpointLoadZero()

baijumeswani commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -454,6 +454,18 @@ def testToyBertCheckpointLoadZero():
+
+        for k,v in rank_state_dict['fp16_param'].items():
+            fp32_name = k.split('_fp16')[0]
+            assert_allclose(v, loaded_state_dict['fp32_param'][fp32_name])
```
going from mixed precision to full precision distributed training is yielding same weights between fp16 pre and fp32 post? This does not allign with my findings in the trainer checkpoint migration matrix excel sheet.

ashbhandare replied:
We can investigate the bugs you have found separately. I have added the script for generating the checkpoints as well in the PR, please feel free to regenerate and run the test, but it passes in the CI as well as locally

baijumeswani replied:
After speaking to @ashbhandare and going through my test scenarios, it seems I might have worked with a corrupt checkpoint file which was causing my test to fail. Some of the error scenarios I reported are now no longer failing. Investigating more on this and will update the matrix as well as the bug I filed.

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,106 @@ static Status AddL2NormNcclAllReduce(
+      view_inputs.push_back(shape_argdef);
+      view_outputs.push_back(view_argdef);
+      view_num++;
+      std::cout<< i << " View:" << view_argdef.name << "\n";
```
older commit.

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -136,23 +205,43 @@ static Status AddViewForParameters(
+      std::cout<< "Test partitoned: " << test_partitioned_weights[i].name << "\n";
+    }
+    weight_views = test_partitioned_weights;
+    // weight_views = AddViewForParameter(graph_defs, weight_argdef, view_shapes);
```
older commit

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -136,23 +205,43 @@ static Status AddViewForParameters(
+  std::vector<ArgDef> weight_views, mixed_precision_weight_views;
   if (opt_config.mixed_precision_weight_arg != nullptr) {
+    //FP32 weight
+    std::vector<ArgDef> test_partitioned_weights = AddPartitionsForParameter(graph, graph_defs, weight_argdef.name, view_shapes);
```
older commit

ashbhandare commented on orttraining/orttraining/python/checkpointing_utils.py:
```diff
@@ -1,143 +1,124 @@
+        self.sharded_params = set()
+
+    def _split_name(self, name):
         name_split = name.split('_view_')
```
This PR completely shards the FP32 weight and adds the same _view_ suffix so that the sharded weight can be identified as such

baijumeswani replied:
If fp32 state names also have ```_view``` in their names, it would break backwards compatibility. Which means, if there are users out there relying on the fp32 weight keynames not having ```_view_```, this change would break their script/code. Do we have any mitigation plan for that?

ashbhandare replied:
This is a breaking change, and since we have experimental support for checkpointing, we do not promise backward compatibility. The user can use an earlier commit if required, although the preferred path would be to migrate to the newer usage.

ashbhandare commented on orttraining/orttraining/python/checkpointing_utils.py:
```diff
@@ -1,143 +1,124 @@
+            weight_name, optimizer_key, view_num, fp16_key = self._split_name(k)
+            if view_num is not None:
+                # parameter is sharded
+                param_name = optimizer_key + weight_name + fp16_key
```
The 'fp16_key' denotes the suffix added to the parameter to declare that it was mixedprecision. Changing the variable name to improve understanding.

baijumeswani replied:
Leaving this here for future reference. An assumption cannot be safely made that if the keyname ends with ```'fp_16'```, it will not be sharded. That is to say that if there is a state key that ends with ```'fp_16'```, it can be sharded.

ashbhandare commented on orttraining/orttraining/python/checkpointing_utils.py:
```diff
@@ -1,143 +1,124 @@
+                    self.aggregate_state_dict[param_name] = v
+            else:
+                if k in self.aggregate_state_dict:
+                    assert (self.aggregate_state_dict[k] == v).all(), "Unsharded params must have the same value"
```
If this happened, This indicates that the checkpoint isn't valid- either due to buggy zero partitioning/training or mismatched checkpoints being loaded together.

ashbhandare commented on orttraining/orttraining/python/training/checkpoint.py:
```diff
@@ -160,6 +159,21 @@ def _get_checkpoint_name(prefix, is_partitioned, world_rank=None, world_size=Non
     return filename
 
 
+def _split_state_dict(state_dict):
```
i am not sure if we wanted to add imports between the old and new APIs.

ashbhandare commented on orttraining/orttraining/python/checkpointing_utils.py:
```diff
@@ -1,143 +1,124 @@
-    def __init__(self, checkpoint_files, clean_state_dict = None):
 
-        assert len(checkpoint_files) > 0, "No checkpoint files passed."
+def split_state_dict(state_dict):
```
it is not. changing

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -214,6 +304,12 @@ static Status ModifyParametersForOptimizerPartitioning(
+    // We only run the optimizer builder for weights belonging to this rank. For weights handled by other ranks,
+    // immediately grab the FP16 weight (if it exists).
+    // ArgDef fp16_or_fp32_weight_argdef = opt_config.fp16_weight_arg != nullptr ? 
+    //     ArgDef(opt_config.fp16_weight_arg->Name(), opt_config.fp16_weight_arg->TypeAsProto()) : weight_argdef;
```
older commit

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,106 @@ static Status AddL2NormNcclAllReduce(
+
+        // Clear the initializer shape, as we are replacing the full weight with a segment of the weight.
+        // NodeArg* initializer_arg = graph.GetNodeArg(initializer_name);
+        // initializer_arg->ClearShape();      
```
older commit

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,106 @@ static Status AddL2NormNcclAllReduce(
+  size_t view_num = 0;
+  for (size_t i = 0; i < shapes.size(); i++) {
+    if (shapes[i].Size() > 0) {
+      std::string partition_name = initializer_name + "_view_" + std::to_string(view_num);
```
changed to get the name from a function call

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,106 @@ static Status AddL2NormNcclAllReduce(
 }
 
-static Status AddViewForParameters(
+static Status AddParameterPartition(
```
since both are valid ways of doing the same thing, not changing to also keep the style consistent with rest of the optimizer graph builder code.

ashbhandare commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -136,23 +205,43 @@ static Status AddViewForParameters(
+    //FP32 weight
+    std::vector<ArgDef> test_partitioned_weights = AddPartitionsForParameter(graph, graph_defs, weight_argdef.name, view_shapes);
+    ORT_ENFORCE(test_partitioned_weights.size() == enabled.size());
+    for (size_t i=0; i < test_partitioned_weights.size(); i++){
```
older commit

baijumeswani reviewed (dismissed):
Looks good.

thiagocrepaldi commented:
> d zero partitioning is conditioned upo
@ashbhandare Are you familiar with the PyTorch flexible API specs? It is a new frontend for ORT which requires all graph initializers to be passed as graph inputs. The initializers come from the original pytorch model and passed into ORT, so the ORT backend is actually stateless in this sense, as it will only compute stuff on top of inputs.

@mrry Do you think this behavior is compatible with `ORTModule` design? Do we intend to support ZeRO on the flexible API?

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -422,8 +421,113 @@ def testToyBertCheckpointBasic():
+
+# checkpointfiles generated using:
+# mpirun -n 4 python3 orttrainer_bert_toy_onnx_ckpt_gen.py
+def testToyBertCheckpointLoadZeroFullprecisionAdam():
```
What is the difference for the previous test? If this is only the optimizer, please use parameterized tests. You can pass optimizer and expected values as parameters and the implementation remain the same

Refer to https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py#L598-L603

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -433,6 +537,10 @@ def testToyBertCheckpointLoadZero():
                                          'device' : {'id' : device},
+                                         'mixed_precision': {
+                                                'enabled': True,
+                                                'loss_scaler': None
```
```suggestion
```
`loss_scaler=None` is not supported

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -454,10 +562,84 @@ def testToyBertCheckpointLoadZero():
 
+# checkpointfiles generated using:
+# mpirun -n 4 python3 orttrainer_bert_toy_onnx_ckpt_gen.py
+def testToyBertCheckpointLoadZeroMixedprecisionAdam():
```
Same as the full precision case. If the difference from the previous test is the optimizer, please parameterize it and keep a single implementation

baijumeswani commented on orttraining/orttraining/python/checkpointing_utils.py:
```diff
@@ -1,143 +1,123 @@
-    def __init__(self, checkpoint_files, clean_state_dict = None):
 
-        assert len(checkpoint_files) > 0, "No checkpoint files passed."
+def _split_state_dict(state_dict):
+    optimizer_keys = ['Moment_1_', 'Moment_2_', 'Update_Count_', 'Step_']
+    split_sd = {'optimizer': {}, 'fp32_param': {}, 'fp16_param': {}}
+    for k, v in state_dict.items():
+        mode = 'fp32_param'
+        for optim_key in optimizer_keys:
+            if k.startswith(optim_key):
+                mode = 'optimizer'
+                break
+        if k.endswith('_fp16'):
+            mode = 'fp16_param'
+        split_sd[mode][k] = v
+    return split_sd
```
For the new design, our state dictionary has separate model and optimizer states. In addition, the fp32 and the fp16 states are also separate for both model and optimizer states. So, this entire function would no longer be needed since all the states are already split.
However, with the current design, we will not be able to completely remove the string matching that is happening in the function ```_split_name(self, name)```. This was the trade off we had to make when considering to save the relationship map and doing string manipulation. We will not be storing the complete relationship map (thereby saving on the save and load times), but will need to continue doing some amount of string matching and manipulation to aggregate the state dictionaries.

thiagocrepaldi replied:
I believe we still can do this without string matching. What is the reason BE cannot provide this information? Why is it not solvable?

ashbhandare replied:
this should be addressed in the new design

baijumeswani replied:
After a discussion with @ashbhandare , I believe we have a design that would work so that we do not do any string comparisons on the front end.

ashbhandare commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -422,8 +421,113 @@ def testToyBertCheckpointBasic():
     for k,v in loaded_sd.items():
         assert torch.all(torch.eq(v, sd[k]))
 
+# checkpointfiles generated using:
+# mpirun -n 4 python3 orttrainer_bert_toy_onnx_ckpt_gen.py
```
added the checkpoint generation in CI that calls save_checkpoint

ashbhandare commented:
> > d zero partitioning is conditioned upo
> 
> @ashbhandare Are you familiar with the PyTorch flexible API specs? It is a new frontend for ORT which requires all graph initializers to be passed as graph inputs. The initializers come from the original pytorch model and passed into ORT, so the ORT backend is actually stateless in this sense, as it will only compute stuff on top of inputs.
> 
> @mrry Do you think this behavior is compatible with `ORTModule` design? Do we intend to support ZeRO on the flexible API?

If the initializers are inputs in the flexible API, will the optimizer be handled by ORT? If not, zero partitioning for stage 1 should not happen within ORT and this change will not be touched. If yes, we could enable the older way of adding a 'View' for the flexible API alone.

baijumeswani commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -422,28 +288,48 @@ def testToyBertCheckpointBasic():
+# state are aggregated and loaded correctly into fp32 and fp16 single-node 
+# trainers respectively, for Adam and Lamb optimizers.
+@pytest.mark.parametrize("optimizer, mixedprecision_enabled, expected_eval_loss", [
+    (optim.LambConfig(), False, [11.011026]),
```
where are these values coming from?

ashbhandare replied:
this is the expected loss given that the checkpoint was loaded correctly

baijumeswani commented on orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:
```diff
@@ -454,10 +340,29 @@ def testToyBertCheckpointLoadZero():
+        
+        for k,v in rank_state_dict['optimizer'].items():
+            if k in loaded_state_dict['optimizer']:
+                assert_allclose(v, loaded_state_dict['optimizer'][k])
```
this assert tests the loading, but it does not test the aggregation. May we have unit test that gives dummy input to the aggregation function, and checks the aggregated output against what we expect.
For example here, we are testing for optimizer states matching. And if they do not match, then we assert that they are sharded. But the real test would be to compare the aggregated states with the expectation. So, do you think we can do anything to test out the aggregation logic?

ashbhandare replied:
aggregation on fp32 weights is indirectly tested by asserting that the eval loss matches the expectation, but adding additional checks to compare aggregation logic separately.

ashbhandare commented:
ORT does not support distributed run on a single device at the moment. Since the CI doesn't have a multi-GPU setup, we cannot add checkpoint generation to the tests for zero. After discussing with @baijumeswani, the solution for now is:
1) added pre-generated checkpoints for the CI running regularly
2) added code to re-generate tests on the e2e tests which run nightly on multi-gpu setup
3) added the 'orttraining_run_bert_pretrain.py ORTBertPretrainTest.test_pretrain_zero' to the e2e pipeline run

When the support required is added, the pre-generated checkpoints must be removed and checkpoints must be generated on the fly.

ashbhandare commented:
Checking in large checkpoint files was causing the android CI to fail. Moving the checkpoint aggregation test to nightly, where checkpoint files can be generated.

thiagocrepaldi commented:
Why checking in checkpoint files? Shouldn' t they be generated and consumed during test time?

ashbhandare commented:
> Why checking in checkpoint files? Shouldn' t they be generated and consumed during test time?

The files being checked in are aggregated reference checkpoint files as the rest of the test code doesn't check optimizer state aggregation. I will try to find an alternate way of verifying optimizer state aggregation

thiagocrepaldi commented:
> > Why checking in checkpoint files? Shouldn' t they be generated and consumed during test time?
> 
> The files being checked in are aggregated reference checkpoint files as the rest of the test code doesn't check optimizer state aggregation. I will try to find an alternate way of verifying optimizer state aggregation

See if that helps: https://github.com/microsoft/onnxruntime/tree/master/orttraining/orttraining/test/python/checkpoint

baijumeswani commented on orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:
```diff
@@ -93,39 +95,97 @@ static Status AddL2NormNcclAllReduce(
+        graph.AddInitializedTensor(initializer_partition);
+
+        //add the modified weight name to get state 
+        updated_weight_names_map.insert({initializer_name, partition_name});
```
one comment here since I don't know the intent behind this statement. Just wanted to make sure that the intent here is not to overwrite the element in the map but rather insert only if the element does not exist in the map, right?

ashbhandare replied:
the first scenario doesnt exist here, but changing in any case

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_distributed_tests.py:
```diff
@@ -14,11 +14,26 @@
     level=logging.DEBUG)
 log = logging.getLogger("DistributedTests")
 
+
```
Refactor into https://github.com/microsoft/onnxruntime/tree/master/orttraining/orttraining/test/python/checkpoint
to keep all checkpoint related tests together

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_checkpoint_aggregation.py:
```diff
@@ -0,0 +1,128 @@
+])
+def testToyBertCheckpointLoadZero(optimizer, mixedprecision_enabled, expected_eval_loss):
+    # Common setup
+    rtol = 1e-03
```
We have been using atol=1e-7 for accuracy. rtol=1e-3 doesn't capture a lot of errors, especially for fp32 scenarios

thiagocrepaldi reviewed (commented):
Overall, the frontend part looks good.

My ask is to refactor it to be in the same folder and reusing the same structure as defined at https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/test/python/orttraining_test_checkpoint.py and https://github.com/microsoft/onnxruntime/tree/master/orttraining/orttraining/test/python/checkpoint 

Also, lets try no check-in *.pt files for this PR. We can generate, aggregate, load and assert them in runtime. This ensures both save and load logic are working. https://github.com/microsoft/onnxruntime/pull/5678 has a lot of pointers on how to achieve that

thiagocrepaldi commented:
/azp run orttraining-linux-gpu-ci-pipeline

thiagocrepaldi commented:
/azp run orttraining-linux-gpu-ci-pipeline

thiagocrepaldi commented on onnxruntime/test/python/onnxruntime_test_ort_trainer.py:
```diff
@@ -686,36 +686,6 @@ def testBertTrainingGradientAccumulation(self):
         assert_allclose(expected_losses, actual_losses, rtol=rtol, err_msg="loss mismatch")
         assert_allclose(expected_eval_loss, actual_eval_loss, rtol=rtol, err_msg="evaluation loss mismatch")
 
-    def testBertCheckpointingBasic(self):
```
Why is this being deleted? This is a valid smoke test for checkpoint

ashbhandare replied:
This test was initially removed as the functionality was covered in more extensive checkpoint test. However, after an offline discussion, this test is required to test the following sceanrio:
"Loading state into a un-initialized trainer(before running eval/train step) In this case, the state is maintained on a temporary variable until eval/step is executed. However, even before that, the user must be able to access state dict values.     

testToyBertCheckpointBasic tests whether the state is there before onnx export

"
